RFC-0001-Agile Coretime.md

```diff
@@ -0,0 +1,291 @@
+
+```rust
+type SaleIndex = u16;
+struct RegionId {
+    core: CoreIndex,  // A `u16`.
+    begin: Timeslice, // A `u16`.
+}
+struct RegionRecord {
+    owner: AccountId,
+    end: Timeslice,
+    price: Option<Balance>,
+    allocation: Option<Vec<ParaId>>, // begins set to `None`
+}
+```
+
+This map functions essentially as a linked list. With one region's `end` field functioning as the next region's key's `begin` field. It is keyyed by the sale index in order to allow the following sale period's Coretime to be manipulated during the `LEADIN_PERIOD` prior to it becoming allocatable.
+
+An additional storage map is maintained to keep the "heads" of this linked list. It is called `NextRegion` and it maps `CoreIndex` to `Timeslice`, to indicate the earliest stored region of the given core.
+
+Notably, if a region is split or transferred, then the `price` is reset to `None`, which has the effect of disallowing a price-increase-capped renewal.
+
+The Broker-chain provides feedback to the Relay-chain on which `ParaId` sets should be serviced on which cores, and does so as they change. Each block, the Broker-chain checks if the period of a `Timeslice` has elapsed. If so, it checks to see if any cores have a newly active `RegionRecord` value in the `Regions` map. If they do, it MUST notify the Relay-chain of the new responsibilities of the relevant `core`. In this case, it MUST remove the item from the `Regions` map and update the `NextRegion` map so it maps the relevant core to the value of removed record's `end` field.
+
+If the `RegionRecord` value for an elapsed `RegionId` has an `allocation` of `None`, then the item is not removed and the Relay-chain is instructed to place the core for instantaneous use.
+
+### Specific functions of the Broker-chain
+
+Several functions of the Broker-chain SHALL be exposed through dispatchables and/or a `nonfungible` trait implementation integrated into XCM:
+
+#### 1. Transfer of ownership
+
+A `transfer(region: RegionId, new_owner: AccountId)` dispatchable SHALL have the effect of altering the current `owner` field of `region` in the `Regions` map from the signed origin to `new_owner`.
+
+An implementation of the `nonfungible` trait SHOULD include equivalent functionality. `RegionId` SHOULD be used for the `AssetInstance` value.
+
+In both cases, the `price` field SHALL become `None`.
+
+#### 2. Split into segments
+
+A `split(region: RegionId, pivot: Timeslice)` dispatchable SHALL have the effect of mutating the Regions entry of key `region` so that the `end` field becomes `pivot` and create a new Regions entry of the same `core` but a `begin` equal to `pivot` whose `RegionRecord` has the same `owner`, `end` and `allocation` fields as the origin value of `region`.
+
+`price` in both records is/becomes `None`.
+
+Also:
+- `owner` field of `region` must the equal to the Signed origin.
```

Comment:
Does it need to be `Signed`, or just something that maps to an `AccountId` (as in `new_owner` within `transfer`)? For example, it could be owned by a collective or parachain. Or it could be signed with a session key corresponding to a collator (in order to claim allocation).

Reply:
Perhaps session key doesn't work if it needs to be possible to authorise transfer.

Reply:
Indeed. The `owner` must be strictly equivalent to be the inner of the `Signed` origin.